corehq/apps/cleanup/management/commands/local_commcare_export.py

```diff
@@ -0,0 +1,231 @@
+    es_query = es_search_by_params(params, project, ['include_archived'])
+    if include_archived:
+        es_query['filter']['and'].append({'or': [
+            {'term': {'doc_type': 'xforminstance'}},
+            {'term': {'doc_type': 'xformarchived'}},
+        ]})
+    else:
+        es_query['filter']['and'].append({'term': {'doc_type': 'xforminstance'}})
+
+    query_set = ElasticAPIQuerySet(
+        payload=es_query,
+        model=ESXFormInstance,
+        es_client=XFormES(project),
+    ).order_by('-received_on')
+    return MockApi(
+        query_set, XFormInstanceResource(), XFormInstanceSerializer()
+    )
+
+
+def _get_mock_api(resource, project, params):
+    if resource == 'case':
+        return _get_case_mock(project, params)
+    elif resource == 'form':
+        return _get_form_mock(project, params)
+    else:
+        raise ValueError("Unknown/unsupported resource type '{}'".format(resource))
+
+
+class LocalCommCareHqClient(object):
+    """
+    Like CommCareHqClient but for a local environment
+    """
+    def __init__(self, url, project, checkpoint_manager=None):
+        self.url = url
+        self.project = project
+        self._checkpoint_manager = checkpoint_manager
+
+    # todo: backoff?
+    # @backoff.on_exception(
+    #     backoff.expo, requests.exceptions.RequestException,
+    #     max_time=300, giveup=is_client_error,
+    #     on_backoff=on_backoff, on_giveup=on_giveup
+    # )
+    def get(self, es_query_set, start, limit, params=None):
+        """
+        """
+        from commcare_export.cli import logger
+        logger.debug("Fetching batch: {}-{}".format(start, limit))
+        # logger.debug("Fetching batch: %s", params)
+        return es_query_set[start:start + limit]
+
+    def iterate(self, resource, paginator, params=None):
+        """
+        Iterates through what the API would have been had it been passed in.
+        """
+        from commcare_export.cli import logger
+
+        # resource is either 'form' or 'case'
+        # params are api params
+        # (e.g. {'limit': 1000, u'type': u'pregnant_mother', 'order_by': 'server_date_modified'})
+        params = dict(params or {})
+        mock_api = _get_mock_api(resource, self.project, params)
+
+        def iterate_resource(resource=resource, params=params):
+            more_to_fetch = True
+            last_batch_ids = set()
+
+            count = 0
+            limit = params['limit']
+            total_count = mock_api.query_set.count()
+            while more_to_fetch:
+                batch = self.get(mock_api.query_set, count, limit, params)
+                batch_list = [mock_api.serialize(obj) for obj in batch]
+                logger.debug('Received {}-{} of {}', count, count + limit, total_count)
+
+                if not batch_list:
+                    more_to_fetch = False
+                else:
+                    for obj in batch_list:
+                        if obj['id'] not in last_batch_ids:
```

Comment:
it is possible since the paginations is date based so there are times when all the docs in a page have the same date (e.g. bulk uploads)

Reply:
Ah! So if `limit` is less than the number of results for the same day, storing only the IDs of the last batch could still result in duplicates? Sounds like date-based pagination is a good idea.

Reply:
I'm not sure how the scenario being described is relevantto docs being skipped. The date-based parameter is typically either immutable (`recieved_on` for forms) or only able to move forwards in time (`server_modified_on` for cases). And the comparison uses a `gte` so I think would not be able to skip? Also it would only be relevant in the checkpoint/restart logic that actually bumps the date, but not this logic, which is just paging through a single ES query...

The "skip" scenario I can envision is maybe if the ES paging order is not consistent for tiebreaking for some reason?

I actually suspect that the above code is not necessary in this version of the export and is more relevant to the API-based approach it was copied from. I can replace it with an exception if we want to test that.